React Native Conflicting peer dependency

Hey, I'm getting a Conflicting peer dependency error after installing @react-three/fiber.
Here's the error
``` 
❯ npm install @react-three/drei
npm error code ERESOLVE
npm error ERESOLVE could not resolve
npm error
npm error While resolving: react-dom@18.3.1
npm error Found: react@18.2.0
npm error node_modules/react
npm error   peer react@"*" from @react-native/virtualized-lists@0.74.84
npm error   node_modules/@react-native/virtualized-lists
npm error     @react-native/virtualized-lists@"0.74.84" from react-native@0.74.2
npm error     node_modules/react-native
npm error       peer react-native@"*" from @react-native/virtualized-lists@0.74.84
npm error       8 more (@react-navigation/elements, ...)
npm error   peer react@"*" from @react-navigation/core@6.4.16
npm error   node_modules/@react-navigation/core
npm error     @react-navigation/core@"^6.4.16" from @react-navigation/native@6.1.17
npm error     node_modules/@react-navigation/native
npm error       peer @react-navigation/native@"^6.0.0" from @react-navigation/elements@1.3.30
npm error       node_modules/@react-navigation/elements
npm error         @react-navigation/elements@"^1.3.30" from @react-navigation/stack@6.3.29
npm error         node_modules/@react-navigation/stack
npm error       2 more (@react-navigation/stack, the root project)
npm error   25 more (@react-navigation/elements, ...)
npm error
npm error Could not resolve dependency:
npm error peer react@"^18.3.1" from react-dom@18.3.1
npm error node_modules/react-dom
npm error   peerOptional react-dom@">=18.0" from @react-three/fiber@8.16.8
npm error   node_modules/@react-three/fiber
npm error     @react-three/fiber@"^8.16.8" from the root project
npm error     2 more (@react-three/drei, @react-spring/three)
npm error   peer react-dom@">=16.13" from react-use-measure@2.1.1
npm error   node_modules/@react-three/fiber/node_modules/react-use-measure
npm error     react-use-measure@"^2.1.1" from @react-three/fiber@8.16.8
npm error     node_modules/@react-three/fiber
npm error       @react-three/fiber@"^8.16.8" from the root project
npm error       2 more (@react-three/drei, @react-spring/three)
npm error   1 more (@react-three/drei)
npm error
npm error Conflicting peer dependency: react@18.3.1
npm error node_modules/react
npm error   peer react@"^18.3.1" from react-dom@18.3.1
npm error   node_modules/react-dom
npm error     peerOptional react-dom@">=18.0" from @react-three/fiber@8.16.8
npm error     node_modules/@react-three/fiber
npm error       @react-three/fiber@"^8.16.8" from the root project
npm error       2 more (@react-three/drei, @react-spring/three)
npm error     peer react-dom@">=16.13" from react-use-measure@2.1.1
npm error     node_modules/@react-three/fiber/node_modules/react-use-measure
npm error       react-use-measure@"^2.1.1" from @react-three/fiber@8.16.8
npm error       node_modules/@react-three/fiber
npm error         @react-three/fiber@"^8.16.8" from the root project
npm error         2 more (@react-three/drei, @react-spring/three)
npm error     1 more (@react-three/drei)
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
```

and this is the package.json

```json
  "dependencies": {
    "@react-navigation/native": "^6.1.17",
    "@react-navigation/stack": "^6.3.29",
    "@react-three/fiber": "^8.16.8",
    "@types/three": "^0.165.0",
    "expo": "^51.0.0",
    "expo-gl": "~14.0.2",
    "react": "18.2.0",
    "react-native": "0.74.2",
    "react-native-gesture-handler": "^2.17.1",
    "react-native-safe-area-context": "^4.10.5",
    "react-native-screens": "^3.32.0",
    "three": "^0.165.0"
  },
```

--- a/example/src/demos/SVGRenderer.tsx
+++ b/example/src/demos/SVGRenderer.tsx
@@ -9,8 +9,7 @@ import {
   events,
   ReconcilerRoot,
 } from '@react-three/fiber'
-import useMeasure, { Options as ResizeOptions } from 'react-use-measure'
-import mergeRefs from 'react-merge-refs'
+import { useMeasure, Options as ResizeOptions } from '../../../packages/fiber/src/web/use-measure'
 import { SVGRenderer } from 'three-stdlib'
 
 function TorusKnot() {
@@ -47,6 +46,7 @@ function Canvas({ children, resize, style, className, ...props }: Props) {
 
   const [bind, size] = useMeasure({ scroll: true, debounce: { scroll: 50, resize: 0 }, ...resize })
   const ref = React.useRef<HTMLDivElement>(null!)
+  React.useImperativeHandle(bind, () => ref.current, [])
   const [gl] = useState(() => new SVGRenderer() as unknown as THREE.WebGLRenderer)
   const root = React.useRef<ReconcilerRoot<HTMLElement>>(null!)
 
@@ -67,7 +67,7 @@ function Canvas({ children, resize, style, className, ...props }: Props) {
 
   return (
     <div
-      ref={mergeRefs([ref, bind])}
+      ref={ref}
       className={className}
       style={{ position: 'relative', width: '100%', height: '100%', overflow: 'hidden', ...style }}
     />

--- a/packages/fiber/package.json
+++ b/packages/fiber/package.json
@@ -43,13 +43,14 @@
   },
   "dependencies": {
     "@babel/runtime": "^7.17.8",
+    "@types/debounce": "^1.2.1",
     "@types/react-reconciler": "^0.26.7",
     "@types/webxr": "*",
     "base64-js": "^1.5.1",
     "buffer": "^6.0.3",
+    "debounce": "^1.2.1",
     "its-fine": "^1.0.6",
     "react-reconciler": "^0.27.0",
-    "react-use-measure": "^2.1.1",
     "scheduler": "^0.21.0",
     "suspend-react": "^0.1.3",
     "zustand": "^3.7.1"

--- a/packages/fiber/src/web/Canvas.tsx
+++ b/packages/fiber/src/web/Canvas.tsx
@@ -1,7 +1,6 @@
 import * as React from 'react'
 import * as THREE from 'three'
-import useMeasure from 'react-use-measure'
-import type { Options as ResizeOptions } from 'react-use-measure'
+import { useMeasure, Options as ResizeOptions } from './use-measure'
 import { useContextBridge, FiberProvider } from 'its-fine'
 import { isRef, SetBlock, Block, ErrorBoundary, useMutableCallback, useIsomorphicLayoutEffect } from '../core/utils'
 import { ReconcilerRoot, extend, createRoot, unmountComponentAtNode, RenderProps } from '../core'

--- a/packages/fiber/src/web/use-measure.ts
+++ b/packages/fiber/src/web/use-measure.ts
@@ -1,0 +1,194 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+import { useEffect, useState, useRef, useMemo } from 'react'
+import createDebounce from 'debounce'
+
+declare type ResizeObserverCallback = (entries: any[], observer: ResizeObserver) => void
+declare class ResizeObserver {
+  constructor(callback: ResizeObserverCallback)
+  observe(target: Element, options?: any): void
+  unobserve(target: Element): void
+  disconnect(): void
+  static toString(): string
+}
+
+export interface RectReadOnly {
+  readonly x: number
+  readonly y: number
+  readonly width: number
+  readonly height: number
+  readonly top: number
+  readonly right: number
+  readonly bottom: number
+  readonly left: number
+  [key: string]: number
+}
+
+type HTMLOrSVGElement = HTMLElement | SVGElement
+
+type Result = [(element: HTMLOrSVGElement | null) => void, RectReadOnly, () => void]
+
+type State = {
+  element: HTMLOrSVGElement | null
+  scrollContainers: HTMLOrSVGElement[] | null
+  resizeObserver: ResizeObserver | null
+  lastBounds: RectReadOnly
+}
+
+export type Options = {
+  debounce?: number | { scroll: number; resize: number }
+  scroll?: boolean
+  polyfill?: { new (cb: ResizeObserverCallback): ResizeObserver }
+  offsetSize?: boolean
+}
+
+export function useMeasure(
+  { debounce, scroll, polyfill, offsetSize }: Options = { debounce: 0, scroll: false, offsetSize: false },
+): Result {
+  const ResizeObserver = polyfill || (typeof window !== 'undefined' && (window as any).ResizeObserver)
+
+  const [bounds, set] = useState<RectReadOnly>({
+    left: 0,
+    top: 0,
+    width: 0,
+    height: 0,
+    bottom: 0,
+    right: 0,
+    x: 0,
+    y: 0,
+  })
+
+  // In test mode
+  if (!ResizeObserver) {
+    // @ts-ignore
+    bounds.width = 1280
+    // @ts-ignore
+    bounds.height = 800
+    return [() => {}, bounds, () => {}]
+  }
+
+  // keep all state in a ref
+  const state = useRef<State>({ element: null, scrollContainers: null, resizeObserver: null, lastBounds: bounds })
+
+  // set actual debounce values early, so effects know if they should react accordingly
+  const scrollDebounce = debounce ? (typeof debounce === 'number' ? debounce : debounce.scroll) : null
+  const resizeDebounce = debounce ? (typeof debounce === 'number' ? debounce : debounce.resize) : null
+
+  // make sure to update state only as long as the component is truly mounted
+  const mounted = useRef(false)
+  useEffect(() => {
+    mounted.current = true
+    return () => void (mounted.current = false)
+  })
+
+  // memoize handlers, so event-listeners know when they should update
+  const [forceRefresh, resizeChange, scrollChange] = useMemo(() => {
+    const callback = () => {
+      if (!state.current.element) return
+      const { left, top, width, height, bottom, right, x, y } =
+        state.current.element.getBoundingClientRect() as unknown as RectReadOnly
+
+      const size = {
+        left,
+        top,
+        width,
+        height,
+        bottom,
+        right,
+        x,
+        y,
+      }
+
+      if (state.current.element instanceof HTMLElement && offsetSize) {
+        size.height = state.current.element.offsetHeight
+        size.width = state.current.element.offsetWidth
+      }
+
+      Object.freeze(size)
+      if (mounted.current && !areBoundsEqual(state.current.lastBounds, size)) set((state.current.lastBounds = size))
+    }
+    return [
+      callback,
+      resizeDebounce ? createDebounce(callback, resizeDebounce) : callback,
+      scrollDebounce ? createDebounce(callback, scrollDebounce) : callback,
+    ]
+  }, [set, offsetSize, scrollDebounce, resizeDebounce])
+
+  // cleanup current scroll-listeners / observers
+  function removeListeners() {
+    if (state.current.scrollContainers) {
+      state.current.scrollContainers.forEach((element) => element.removeEventListener('scroll', scrollChange, true))
+      state.current.scrollContainers = null
+    }
+
+    if (state.current.resizeObserver) {
+      state.current.resizeObserver.disconnect()
+      state.current.resizeObserver = null
+    }
+  }
+
+  // add scroll-listeners / observers
+  function addListeners() {
+    if (!state.current.element) return
+    state.current.resizeObserver = new ResizeObserver(scrollChange)
+    state.current.resizeObserver!.observe(state.current.element)
+    if (scroll && state.current.scrollContainers) {
+      state.current.scrollContainers.forEach((scrollContainer) =>
+        scrollContainer.addEventListener('scroll', scrollChange, { capture: true, passive: true }),
+      )
+    }
+  }
+
+  // the ref we expose to the user
+  const ref = (node: HTMLOrSVGElement | null) => {
+    if (!node || node === state.current.element) return
+    removeListeners()
+    state.current.element = node
+    state.current.scrollContainers = findScrollContainers(node)
+    addListeners()
+  }
+
+  // add general event listeners
+  useOnWindowScroll(scrollChange, Boolean(scroll))
+  useOnWindowResize(resizeChange)
+
+  // respond to changes that are relevant for the listeners
+  useEffect(() => {
+    removeListeners()
+    addListeners()
+  }, [scroll, scrollChange, resizeChange])
+
+  // remove all listeners when the components unmounts
+  useEffect(() => removeListeners, [])
+  return [ref, bounds, forceRefresh]
+}
+
+// Adds native resize listener to window
+function useOnWindowResize(onWindowResize: (event: Event) => void) {
+  useEffect(() => {
+    const cb = onWindowResize
+    window.addEventListener('resize', cb)
+    return () => void window.removeEventListener('resize', cb)
+  }, [onWindowResize])
+}
+function useOnWindowScroll(onScroll: () => void, enabled: boolean) {
+  useEffect(() => {
+    if (enabled) {
+      const cb = onScroll
+      window.addEventListener('scroll', cb, { capture: true, passive: true })
+      return () => void window.removeEventListener('scroll', cb, true)
+    }
+  }, [onScroll, enabled])
+}
+
+// Returns a list of scroll offsets
+function findScrollContainers(element: HTMLOrSVGElement | null): HTMLOrSVGElement[] {
+  const result: HTMLOrSVGElement[] = []
+  if (!element || element === document.body) return result
+  const { overflow, overflowX, overflowY } = window.getComputedStyle(element)
+  if ([overflow, overflowX, overflowY].some((prop) => prop === 'auto' || prop === 'scroll')) result.push(element)
+  return [...result, ...findScrollContainers(element.parentElement)]
+}
+
+// Checks if element boundaries are equal
+const keys: (keyof RectReadOnly)[] = ['x', 'y', 'top', 'bottom', 'left', 'right', 'width', 'height']
+const areBoundsEqual = (a: RectReadOnly, b: RectReadOnly): boolean => keys.every((key) => a[key] === b[key])

